File: doc/porting/JellyItem.cs
using Assets.Scripts.CasualTools.Common.Pooling;
using Assets.Scripts.GamePlayScene.Mechanics.Animations;
using Assets.Scripts.Utils;
using System;
using UnityEngine;

namespace Assets.Scripts.GamePlayScene.Mechanics.Items
{
	public class JellyItem : SpriteBasedItem
	{
		private GameObject _itemToBeDestoyedAtAnimationEnd;

		private JellyItem _jellyToBeEnabled;

		public Transform JellyAnimationOwner;

		private JellyAnimatorWaiter _jellyAnimatorWaiter;

		private bool _creatingJelly;

		private bool _destroyed;

		private JellyItem _newJelly;

		public JellyItem SpreadedBy;

		public override int GetScore()
		{
			return 500;
		}

		public override void PlayExplodeAudio()
		{
			AudioManager.Play(AudioTag.JellyBreak, PlayMode.Frame, null, 1f);
		}

		public void PlayJellySpread(Cell cellSelected)
		{
			if (this._creatingJelly || this._destroyed || cellSelected == null)
			{
				return;
			}
			AudioManager.Play(AudioTag.JellySpread, PlayMode.Frame, null, 1f);
			this._jellyAnimatorWaiter = this.CurrentLevel.CurrentLevelBuilder.JellyAnimatorPrefab.Spawn(this.JellyAnimationOwner.get_position()).GetComponent<JellyAnimatorWaiter>();
			this._jellyAnimatorWaiter.MasterJelly = this;
			int num = 0;
			GameObject prefab = ParticlePool.Instance.JellyCreationSideParticle;
			if (cellSelected.Y == this.CurrentCell.Y)
			{
				num = ((cellSelected.X >= this.CurrentCell.X) ? 3 : 1);
				this._jellyAnimatorWaiter.DisplaySortNo = this.CurrentSpriteRenderer.get_sortingOrder() + 5;
			}
			else if (cellSelected.X == this.CurrentCell.X)
			{
				if (cellSelected.Y < this.CurrentCell.Y)
				{
					num = 4;
					prefab = ParticlePool.Instance.JellyCreationDownParticle;
					this._jellyAnimatorWaiter.DisplaySortNo = this.CurrentSpriteRenderer.get_sortingOrder() - 1;
				}
				else
				{
					num = 2;
					prefab = ParticlePool.Instance.JellyCreationUpParticle;
					this._jellyAnimatorWaiter.DisplaySortNo = this.CurrentSpriteRenderer.get_sortingOrder() + 105;
				}
			}
			this._jellyAnimatorWaiter.HideIt();
			if (num == 0)
			{
				this._jellyAnimatorWaiter.MasterJelly = null;
				this._jellyAnimatorWaiter.get_gameObject().Recycle();
				return;
			}
			this._creatingJelly = true;
			this._jellyAnimatorWaiter.SpreadAnimator.SetInteger("DIRECTION", num);
			if (cellSelected.HasItem)
			{
				Item currentItem = cellSelected.CurrentItem;
				currentItem.CurrentCell = null;
				this._itemToBeDestoyedAtAnimationEnd = currentItem.get_gameObject();
			}
			this._newJelly = (this.CurrentLevel.CurrentLevelBuilder.CreateItem(new ItemDescription(ItemType.Jelly, TiledEditorId.Jelly, null)) as JellyItem);
			this._newJelly.get_transform().set_position(cellSelected.get_transform().get_position());
			this._newJelly.CurrentCell = cellSelected;
			this._newJelly.RelativeContainer.SetActive(false);
			this._newJelly.SpreadedBy = this;
			ParticlePlayer particlePlayer = ParticlePool.Instance.PlayParticle(prefab, this._newJelly.get_transform().get_position(), false);
			particlePlayer.get_transform().set_localScale((num != 3) ? Vector3.get_one() : new Vector3(-1f, 1f, 1f));
			particlePlayer.Play();
			this._jellyToBeEnabled = this._newJelly;
		}

		public override bool CanShadowDrop()
		{
			return true;
		}

		public void AnimationEnded()
		{
			if (this._newJelly != null)
			{
				this._newJelly.SpreadedBy = null;
				if (this._newJelly.CurrentCell != null)
				{
					this._newJelly.get_transform().set_position(this._newJelly.CurrentCell.get_transform().get_position());
				}
			}
			if (this._jellyAnimatorWaiter != null)
			{
				this._jellyAnimatorWaiter.SpreadAnimator.SetInteger("DIRECTION", 5);
				this._jellyAnimatorWaiter.MasterJelly = null;
				this._jellyAnimatorWaiter.HideIt();
				this._jellyAnimatorWaiter.get_gameObject().Recycle();
			}
			if (this._itemToBeDestoyedAtAnimationEnd != null)
			{
				this._itemToBeDestoyedAtAnimationEnd.GetComponent<Item>().RemoveSelf();
			}
			if (this._jellyToBeEnabled != null)
			{
				this._jellyToBeEnabled.RelativeContainer.SetActive(true);
			}
			this._jellyToBeEnabled = null;
			this._itemToBeDestoyedAtAnimationEnd = null;
			this._jellyAnimatorWaiter = null;
			this._creatingJelly = false;
		}

		public override void OnItemCreated(Level currentLevel)
		{
			base.OnItemCreated(currentLevel);
			this.CurrentLevel.JellyItemManager.Add(this);
		}

		public override void OnBeforeRemove()
		{
			if (this.SpreadedBy != null)
			{
				this.SpreadedBy.AnimationEnded();
			}
			this.AnimationEnded();
			base.OnBeforeRemove();
			this.CurrentLevel.JellyItemManager.Remove(this);
			this._destroyed = true;
		}

		public override void AfterPoolInit()
		{
			base.AfterPoolInit();
			this._destroyed = false;
		}

		public override void ReactToUpperFall()
		{
			FallListener fallListener = base.GetFallListener();
			fallListener.EndFall(-1f);
		}

		protected override void PlayOnExplodeStartedAnimation()
		{
			if (ParticlePool.Instance != null)
			{
				ParticlePool.Instance.PlayParticle(ParticlePool.Instance.JellyExplodeParticle, base.get_transform().get_position(), true);
			}
		}

		public override bool CanExplodeDueToExplodeInNeigbour(MatchType sourceType)
		{
			return true;
		}

		public override bool CanFall()
		{
			return false;
		}

		public override bool TryExplode()
		{
			bool flag = base.TryExplode();
			if (flag)
			{
				this.CurrentLevel.JellyItemManager.InformJellyExplode();
			}
			return flag;
		}

		public override ItemType GetItemType()
		{
			return ItemType.Jelly;
		}
	}
}

File: classes/game/item/JellyItem.cpp
#include "JellyItem.h"
#include "JellyAnimatorWaiter.h"
#include "../Cell.h"
#include "../Level.h"
#include "../LevelBuilder.h"
#include "../Animation.h"
#include "../JellyManager.h"
#include "../ParticleLibrary.h"
#include "../ImageLibrary.h"

void JellyItem::SetupItem() {
	ImageLibrary::CopySprite(this->CurrentSpriteRenderer, IMGLIB->Jelly);
}

void JellyItem::RemoveSelf() {
	this->OnBeforeRemove();
	//base.get_gameObject().Recycle();
	this->CurrentSpriteRenderer->removeFromParentAndCleanup(true);
	//delete this;
}

void JellyItem::PlayExplodeAudio() {
//	AudioManager.Play(AudioTag.JellyBreak, PlayMode.Frame, nullptr, 1f);
}

void JellyItem::PlayJellySpread(Cell* cellSelected) {
	if (this->_creatingJelly || this->_destroyed || cellSelected == nullptr) {
		return;
	}
//	AudioManager.Play(AudioTag.JellySpread, PlayMode.Frame, nullptr, 1f);
	this->_jellyAnimatorWaiter = new JellyAnimatorWaiter();//this->CurrentLevel->CurrentLevelBuilder->JellyAnimatorPrefab.Spawn(this->JellyAnimationOwner.get_position()).GetComponent<JellyAnimatorWaiter>();
	this->_jellyAnimatorWaiter->MasterJelly = this;
	int num = 0;
//	cocos2d::Node* prefab = ParticlePool.Instance.JellyCreationSideParticle;
	if (cellSelected->Y == this->get_CurrentCell()->Y) {
		num = ((cellSelected->X >= this->get_CurrentCell()->X) ? 3 : 1);
		this->_jellyAnimatorWaiter->DisplaySortNo = this->CurrentSpriteRenderer->getLocalZOrder() + 5;
	}
	else if (cellSelected->X == this->get_CurrentCell()->X) {
		if (cellSelected->Y < this->get_CurrentCell()->Y) {
			num = 4;
//			prefab = ParticlePool.Instance.JellyCreationDownParticle;
			this->_jellyAnimatorWaiter->DisplaySortNo = this->CurrentSpriteRenderer->getLocalZOrder() - 1;
		}
		else {
			num = 2;
//			prefab = ParticlePool.Instance.JellyCreationUpParticle;
			this->_jellyAnimatorWaiter->DisplaySortNo = this->CurrentSpriteRenderer->getLocalZOrder() + 105;
		}
	}
	this->_jellyAnimatorWaiter->HideIt();
	if (num == 0) {
		this->_jellyAnimatorWaiter->MasterJelly = nullptr;
//		this->_jellyAnimatorWaiter->get_gameObject().Recycle();
		return;
	}
	this->_creatingJelly = true;
//	this->_jellyAnimatorWaiter->SpreadAnimator.SetInteger("DIRECTION", num);
	if (cellSelected->get_HasItem()) {
		Item* currentItem = cellSelected->get_CurrentItem();
		currentItem->set_CurrentCell(nullptr);
		this->_itemToBeDestoyedAtAnimationEnd =currentItem;//currentItem.get_gameObject();
	}
	this->_newJelly = dynamic_cast<JellyItem*>(this->CurrentLevel->CurrentLevelBuilder->CreateItem(new ItemDescription(ItemType::Jelly, TiledEditorId::Jelly, nullptr)));
//	this->_newJelly->get_transform().set_position(cellSelected->get_transform().get_position());
	LevelBuilder::ActiveLevel->CurrentLevelBuilder->setCellItemSpritePosition(cellSelected, this->_newJelly);
	this->_newJelly->set_CurrentCell(cellSelected);
//	this->_newJelly->RelativeContainer.SetActive(false);
	this->_newJelly->SpreadedBy = this;
//	ParticlePlayer particlePlayer = ParticlePool.Instance.PlayParticle(prefab, this->_newJelly->get_transform().get_position(), false);
//	particlePlayer.get_transform().set_localScale((num != 3) ? Vector3.get_one() : new Vector3(-1f, 1f, 1f));
//	particlePlayer.Play();

	this->_newJelly->CurrentSpriteRenderer->setScale(0);
	CurrentSpriteRenderer->runAction(cocos2d::Sequence::create(
		cocos2d::TargetedAction::create(this->_newJelly->CurrentSpriteRenderer,cocos2d::EaseBackOut::create(cocos2d::ScaleTo::create(0.5f,1.0f))),
		cocos2d::CallFunc::create([this]() {this->_jellyAnimatorWaiter->InformAnimationEnded(); }),
		nullptr));


	this->_jellyToBeEnabled = this->_newJelly;
}

void JellyItem::AnimationEnded() {
	if (this->_newJelly != nullptr) {
		this->_newJelly->SpreadedBy = nullptr;
		if (this->_newJelly->get_CurrentCell() != nullptr) {
//			this->_newJelly->get_transform().set_position(this->_newJelly->get_CurrentCell()->get_transform().get_position());
			this->_newJelly->CurrentSpriteRenderer->setPosition(this->_newJelly->get_CurrentCell()->GetFrozenPosition());
		}
	}
	if (this->_jellyAnimatorWaiter != nullptr) {
//		this->_jellyAnimatorWaiter->SpreadAnimator.SetInteger("DIRECTION", 5);
		this->_jellyAnimatorWaiter->MasterJelly = nullptr;
		this->_jellyAnimatorWaiter->HideIt();
//		this->_jellyAnimatorWaiter->get_gameObject().Recycle();
	}
	if (this->_itemToBeDestoyedAtAnimationEnd != nullptr) {
//		this->_itemToBeDestoyedAtAnimationEnd->GetComponent<Item>().RemoveSelf();
		this->_itemToBeDestoyedAtAnimationEnd->RemoveSelf();
	}
	if (this->_jellyToBeEnabled != nullptr) {
//		this->_jellyToBeEnabled->RelativeContainer.SetActive(true);
	}
	this->_jellyToBeEnabled = nullptr;
	this->_itemToBeDestoyedAtAnimationEnd = nullptr;
	this->_jellyAnimatorWaiter = nullptr;
	this->_creatingJelly = false;
}

void JellyItem::OnItemCreated(Level* currentLevel) {
	Item::OnItemCreated(currentLevel);
	this->CurrentLevel->JellyItemManager->Add(this);
}

void JellyItem::OnBeforeRemove() {
	if (this->SpreadedBy != nullptr) {
		this->SpreadedBy->AnimationEnded();
	}
	this->AnimationEnded();
	Item::OnBeforeRemove();
	this->CurrentLevel->JellyItemManager->Remove(this);
	this->_destroyed = true;
}

void JellyItem::AfterPoolInit() {
	Item::AfterPoolInit();
	this->_destroyed = false;
}

void JellyItem::ReactToUpperFall() {
	FallListener* fallListener = Item::GetFallListener();
	if(fallListener!=nullptr)
		fallListener->EndFall(-1.0f);
}

void JellyItem::PlayOnExplodeStartedAnimation() {
//	if (ParticlePool.Instance != nullptr)
//	{
//		ParticlePool.Instance.PlayParticle(ParticlePool.Instance.JellyExplodeParticle, Item::get_transform().get_position(), true);
//	}
	PTCLLIB->explodeJelly(CurrentSpriteRenderer->getPosition());
}

bool JellyItem::TryExplode() {
	bool flag = Item::TryExplode();
	if (flag) {
		this->CurrentLevel->JellyItemManager->InformJellyExplode();
	}
	return flag;
}
